middleware/cancel/cancel.go

```diff
@@ -12,50 +12,49 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Cancel ensures that contexts are canceled when their associated tasks are completed.
-// There are two parts of this middleware:
-// 1. Tying Provider.Cancel to all associated contexts.
-// 2. Applying timeout information when available.
+// The `cancel` package provides a middle-ware that ties the Cancel gRPC call from Pulumi
+// to Go's `context.Context` cancellation system.
+//
+// Wrapping a provider in `cancel.Wrap` ensures 2 things:
+//
+// 1. When a resource operation times out, the associated context is canceled.
+//
+// 2. When `Cancel` is called, all outstanding gRPC methods have their associated contexts
+// canceled.
+//
+// A `cancel.Wrap`ed provider will still call the `Cancel` method on the underlying
+// provider. If NotImplemented is returned, it will be swallowed.
```

Comment:
I added some docs clarifying what `cancel` *currently* does. This behavior is unchanged before and after this PR. The PR only prevents panics on concurrent creations and cancelations.

Since this PR fixes a P1 (panic) and doesn't change the design of `cancel.Wrap`, I suggest that we merge as is unless there are suggestions on how to better implement the existing semantic.

I can tell that there is disagreement on the *desired* behavior of `cancel.Wrap`. I'm happy to have that discussion, but I think it should be a separate discussion, outside the scope of this PR.

CC @blampe @t0yv0

Reply:
I opened https://github.com/pulumi/pulumi-go-provider/issues/208 to track a design discussion on what `cancel` should do.

Reply:
Sounds good to me!